src/lib/operators/join_index.cpp

```diff
@@ -127,16 +127,25 @@ void JoinIndex::_perform_join() {
           // clang-format off
           if constexpr (neither_is_string_column || both_are_string_columns) {
               auto iterable_left = create_iterable_from_column<LeftType>(typed_left_column);
-              auto iterable_right = create_iterable_from_column<RightType>(typed_right_column);
-
-              iterable_left.with_iterators([&](auto left_it, auto left_end) {
-                  iterable_right.with_iterators([&](auto right_it, auto right_end) {
-                      with_comparator(_predicate_condition, [&](auto comparator) {
-                          this->_join_two_columns(comparator, left_it, left_end, right_it, right_end, chunk_id_left,
-                                                  chunk_id_right, index);
-                      });
-                  });
-              });
+
+              if (index != nullptr) {
+                // utilize index for join
+                iterable_left.with_iterators([&](auto left_it, auto left_end) {
+                    this->_join_two_columns_using_index(left_it, left_end, chunk_id_left, chunk_id_right, index);
+                });
```

Comment:
This does still not solve the problem. This `index != nullptr` block needs to be outside of the second `resolve_column_and_data_type`